services/motion/builtin/move_request.go

```diff
@@ -38,17 +36,16 @@ func (mr *moveRequest) plan(ctx context.Context) (motionplan.Plan, error) {
 		inputs = inputs[:2]
 	}
 	mr.planRequest.StartConfiguration = map[string][]referenceframe.Input{mr.kinematicBase.Kinematics().Name(): inputs}
-	return motionplan.PlanMotion(ctx, mr.planRequest)
-}
-
-func (mr *moveRequest) execute(ctx context.Context, plan motionplan.Plan) moveResponse {
-	waypoints, err := plan.GetFrameSteps(mr.kinematicBase.Kinematics().Name())
+	plan, err := motionplan.PlanMotion(ctx, mr.planRequest)
 	if err != nil {
-		return moveResponse{err: err}
+		return nil, err
 	}
+	return plan.GetFrameSteps(mr.kinematicBase.Kinematics().Name())
+}
 
+func (mr *moveRequest) execute(ctx context.Context, waypoints [][]referenceframe.Input, waypointIndex *atomic.Int32) moveResponse {
 	// Iterate through the list of waypoints and issue a command to move to each
-	for i := 1; i < len(waypoints); i++ {
+	for i := waypointIndex.Load(); i < int32(len(waypoints)); i = waypointIndex.Add(1) {
```

Comment:
Is this the only function that adds to the waypoint? 

If so could I recommend splitting out the loop bookkeeping & the side effect of incrementing the atomic counter (which is visible to other go routines) by putting the atomic increment at the boom of the loop. I don't think developers are used to expecting loop counter increments to have side effects that affect other goroutines:

```suggestion
	for i := waypointIndex.Load();) i < int32(len(waypoints)); i++ {
	...
	waypointIndex.Add(1)
	}
```

Reply:
I personally find this less readable.  I will add a comment in the function description noting that the waypoint index will be mutated